### PR TITLE
Improve auto-atomicity for Regex loops

### LINKED ISSUE #62449: Improve Regex support for automatically making loops atomic
In .NET 5, we added some support for automatically making some loops atomic.  For example, given the expression `a+b+c+`, we'll convert that into the equivalent of `(?>a+)(?>b+)(?>c+)`: the `c+` is at the end of the expression so nothing can backtrack into it, the `b+` is always followed by a `c` and there's nothing `b+` can give back that would match `c`, and similarly the `a+` is always followed by a `b` and there's nothing `a+` can give back that would match `b`.  However, tweaking the expression slightly to `a*b*c*`, we'll convert that into the equivalent of `a*b*(?>c*)`: while we still treat the `c*` as atomic, when analyzing the `a*` and `b*` loops and comparing them against what's guaranteed to come next, we see that the subsequent loop has a 0-lower-bound and give up, even though with a bit more analysis we could see that it's still valid to make all of these atomic.  The more we can make atomic, the better, not just for performance but also for readability of code in the source generator.

### PR DESCRIPTION
**Pull request by stephentoub**
- As part of recognizing loops that could be made atomic, we currently only look at the node guaranteed to come immediately after it, e.g. in `a+b+c+`, when analyzing `a+` we only look at the `b+`.  That's fine in many cases, but if our expression was instead e.g. `a*b*c*`, we couldn't make `a*` atomic because we wouldn't be guaranteed it'd be followed by a `b`, and we wouldn't look at the subsequent nodes to determine what else might follow.  When that situation occurs, we can now walk the tree to find the next node, and repeatedly do so as much as is needed in order to find the next node, enabling us to make all of those loops atomic.
- We currently perform the auto-atomicity reduction for single-char loops as part of reducing concatenations.  But this means we're performing that transformation at a time when the tree isn't fully formed, which means we can't necessarily reach out of a concatenation to what comes after it and use that to influence the atomicity.  It also means we might repeat the analysis unnecessarily multiple times, because it's done as part of parenting a node.  We can instead do that optimization as part of the final optimization phase, after the tree is already created.
- When we find a lazy loop at the end of an expression, we can treat it as atomic, which means we can lower its max bound to its min bound.
- When we encounter a loop (lazy or greedy) at the end of an expression, if it has a max bound of 1 (i.e. it's optional), we can proceed to eliminate ending backtracking with its child regardless of whether the child could be made atomic with the start of the loop, since it won't ever repeat more than once.

Fixes https://github.com/dotnet/runtime/issues/62449

### CODE REVIEW

File: src/libraries/System.Text.RegularExpressions/tests/RegexReductionTests.cs [hunk 1]
```diff
@@ -456,7 +468,7 @@ public void PatternsReduceIdentically(string pattern1, string pattern2)
         [InlineData("(?i:abcd)|abcd", "abcd|abcd")]
         [InlineData("abcd|(?i:abcd)", "abcd|abcd")]
         // Not applying auto-atomicity
```
Comment (danmoseley) [inline src/libraries/System.Text.RegularExpressions/tests/RegexReductionTests.cs:470]: does this comment apply to the lines below? as it appears that it is applying auto-atomicity.

Comment (danmoseley) [reply src/libraries/System.Text.RegularExpressions/tests/RegexReductionTests.cs:470]: worth keeping the `[InlineData("a*b*", "(?>a*)b*")]` case? I guess I'm not inclined to reason too much about whether something is necessary to add more coverage or not so long as it's not self evidently equivalent to another..

Comment (danmoseley) [reply src/libraries/System.Text.RegularExpressions/tests/RegexReductionTests.cs:470]: oh disregard that, I see it above.

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexNode.cs [hunk 2]
```diff
@@ -491,11 +491,26 @@ private void EliminateEndingBacktracking()
                         node = node.Child(0);
                         continue;
 
-                    // For Loop, we search to see if there's a viable last expression, and iff there
-                    // is we recur into processing it.
+                    // For {Lazy}Loop, we search to see if there's a viable last expression, and iff there
+                    // is we recur into processing it.  Also, as with the single-char lazy loops, LazyLoop
+                    // can have its max iteration count dropped to its min iteration count, as there's no
+                    // reason for it to match more than the minimal at the end; that in turn makes it a
+                    // repeater, which results in better code generation.
```
Comment (danmoseley) [inline src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexNode.cs:498]: maybe add example of lazy loop case also?

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexNode.cs [hunk 3]
```diff
@@ -1823,110 +1854,189 @@ private RegexNode ReducePrevent()
...
+                            // in and we can declare success.
                             return true;
+
+                        default:
+                            // Anything else, we don't know what to do, so we have to assume it could conflict with the loop.
+                            return false;
                     }
+
                     break;
+                }
```
Comment (danmoseley) [inline src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexNode.cs:2038]: seems like there should be a return false here. I must be misreading.

File: src/libraries/System.Text.RegularExpressions/tests/RegexReductionTests.cs (hunk 1 above)
Comment (stephentoub) [reply src/libraries/System.Text.RegularExpressions/tests/RegexReductionTests.cs:470]: > as it appears that it is applying auto-atomicity.

This test validates they're _not_ the same.

File: src/libraries/System.Text.RegularExpressions/tests/RegexReductionTests.cs [hunk 4]
```diff
@@ -373,12 +373,24 @@ private static int GetMinRequiredLength(Regex r)
...
-        [InlineData("a*b+", "(?>a*)b+")]
-        [InlineData("a*b{3,4}", "(?>a*)b{3,4}")]
+        [InlineData("a*b+", "(?>a*)(?>b+)")]
+        [InlineData("a*b*", "(?>a*)(?>b*)")]
+        [InlineData("a*b+c*", "(?>a*)(?>b+)(?>c*)")]
+        [InlineData("a*b*c*", "(?>a*)(?>b*)(?>c*)")]
+        [InlineData("a*b*c*|d*[ef]*", "(?>a*)(?>b*)(?>c*)|(?>d*)(?>[ef]*)")]
+        [InlineData("(a*)(b*)(c*)", "((?>a*))((?>b*))((?>c*))")]
+        [InlineData("a*b{3,4}", "(?>a*)(?>b{3,4})")]
+        [InlineData("[ab]*[^a]*", "[ab]*(?>[^a]*)")]
```
Comment (danmoseley) [inline src/libraries/System.Text.RegularExpressions/tests/RegexReductionTests.cs:383]: also 
`[InlineData("[aa]*[^a]*", "(?>[aa]*)(?>[^a]*)")]` ? (just following along)

File: src/libraries/System.Text.RegularExpressions/tests/RegexReductionTests.cs [hunk 5]
```diff
@@ -373,12 +373,24 @@ private static int GetMinRequiredLength(Regex r)
...
+        [InlineData("a??", "")]
```
Comment (danmoseley) [inline src/libraries/System.Text.RegularExpressions/tests/RegexReductionTests.cs:384]: `[InlineData("(abc)??", "")` might be worth adding too? I don't see that higher up.

**Review by danmoseley — APPROVED** [review]
LGTM. Largely relying on the tests here.

Would it be interesting to add to your "real world corpus" the "reduced" versions of each one?  Essentially, the stringified output of the result after the optimization process. I assume that stringified output is fairly stable-ish, even though code maintenance could potentially change to an equivalent stringified output. We already have the stringifying code for debugging. 

Then, when we further improve the optimization, we could spot check any diffs in the reduced result for each one, to see whether there's either unexpected regressions, or it suggests interesting more unit tests. Then we update the baseline. (This need not be run as an automated test, only manual, to eliminate the need to make synchronized baseline updates.)

File: src/libraries/System.Text.RegularExpressions/tests/RegexReductionTests.cs (hunk 1 above)
Comment (danmoseley) [reply src/libraries/System.Text.RegularExpressions/tests/RegexReductionTests.cs:470]: Oops I looked above at the method name, but of course it's below for [InlineData], I did that before..

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexNode.cs (hunk 3 above)
Comment (stephentoub) [reply src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexNode.cs:2038]: If we get here, we need to loop around the outermost loop again with the new subsequent we just found.

**Comment by stephentoub** [comment]
From our corpus, this shows before/after total numbers of each kind of node.  You can see the increase in atomic single-char loops in particular (and corresponding decrease in the non-atomic ones):
image

File: src/libraries/System.Text.RegularExpressions/tests/RegexReductionTests.cs (hunk 5 above)
Comment (stephentoub) [reply src/libraries/System.Text.RegularExpressions/tests/RegexReductionTests.cs:384]: Functionally with the source generator / compiler you'll end up with empty, but we don't actually get that today after parsing / reducing, due to the structure of the optimization that runs.  We end up with the equivalent of `(abc){0}?`, but the nature of how this test is verifying things means I can't actually write that.  Opportunities for another day...

**Comment by stephentoub** [comment]
> Would it be interesting to add to your "real world corpus" the "reduced" versions of each one?

I'll leave that up to @joperezr for https://github.com/dotnet/runtime/issues/63300.  It'd need some thought.

**Review by joperezr — APPROVED** [review]

**Pull request merged**